### ISSUE TITLE
Table creations fail due to large index key size

### ISSUE THREAD

**Issue by kitchin** [wp-statistics/wp-statistics#340]
Thanks for your work on 13.0.3, but I'm seeing a problem in `error_log` creating the tables `wp_statistics_visitor`, `wp_statistics_pages` and `wp_statistics_historical`. 

We are on CentOS, PHP 7.3, Mariadb 10.3.

The tables are not created.

Here is the last error:

`WordPress database error Specified key was too long; max key length is 1000 bytes for query 
CREATE TABLE wp_statistics_historical (
	ID bigint(20) NOT NULL AUTO_INCREMENT,
	category varchar(25) NOT NULL,
	page_id bigint(20) NOT NULL,
	uri varchar(255) NOT NULL,
	value bigint(20) NOT NULL,
	PRIMARY KEY  (ID),
	KEY category (category),
	UNIQUE KEY page_id (page_id),
	UNIQUE KEY uri (uri)
) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci made by activate_plugin, do_action('activate_wp-statistics/wp-statistics.php'), WP_Hook->do_action, WP_Hook->apply_filters, WP_Statistics::install, WP_STATISTICS\Install->install, WP_STATISTICS\Install::create_table, WP_STATISTICS\Install::table_sql, dbDelta`

**Comment by mostafasoufi** [comment]
Thanks for reporting, this issue should be fixed in the mentioned commit, could you please make sure that works fine for you? @kitchin

**Comment by kitchin** [comment]
commit e67b0ec88d79dc55decf813c519eeec4c066661a

I'll do some testing, but here's what I see so far. WP core does it a different way, with more emphasis on limiting index size than column size. Both the plugin and core do use a mix , but core is more definitive and clear about what's going on.

`wordpress-5.5.3/wp-admin/includes/schema.php`
The `$max_index_length` calculated is used for creating default tables `wp_termmeta`, `wp_commentmeta`, `wp_postmeta`, `wp_posts`, `wp_usermeta`, and, in WPMU, `wp_blogs` and `wp-site_meta`. Here is the first one:

```
	/*
	 * Indexes have a maximum size of 767 bytes. Historically, we haven't need to be concerned about that.
	 * As of 4.2, however, we moved to utf8mb4, which uses 4 bytes per character. This means that an index which
	 * used to have room for floor(767/3) = 255 characters, now only has room for floor(767/4) = 191 characters.
	 */
	$max_index_length = 191;

	// Blog-specific tables.
	$blog_tables = "CREATE TABLE $wpdb->termmeta (
	meta_id bigint(20) unsigned NOT NULL auto_increment,
	term_id bigint(20) unsigned NOT NULL default '0',
	meta_key varchar(255) default NULL,
	meta_value longtext,
	PRIMARY KEY  (meta_id),
	KEY term_id (term_id),
	KEY meta_key (meta_key($max_index_length))
) $charset_collate;
CREATE TABLE $wpdb->terms (
 term_id bigint(20) unsigned NOT NULL auto_increment,
 name varchar(200) NOT NULL default '',
 slug varchar(200) NOT NULL default '',
 term_group bigint(10) NOT NULL default 0,
 PRIMARY KEY  (term_id),
 KEY slug (slug($max_index_length)),
 KEY name (name($max_index_length))
) $charset_collate;
```

**Comment by kitchin** [comment]
OK! The current code (will be 13.0.4?) creates the tables without error.

I did not see an error truncating existing data with `WP_DEBUG` true nor in `wp-admin/error_log`, so that's good.  An old thread on Mysql in stack says it can: https://stackoverflow.com/questions/9611100/mysql-how-to-truncate-the-length-of-a-field

WP core `dbDelta()`, used in this plugin, applies neither mitigation `ALTER IGNORE` nor `LEFT(...)` to prevent such an error. But I saw no problem on my test site after populating a table with max characters and then activating the plugin.
